server/api/impl/admin/synchronize.go

```diff
@@ -0,0 +1,186 @@
+
+	"github.com/mattermost/mattermost-plugin-apps/server/api"
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/pkg/errors"
+)
+
+// appsMappingsEnvVarName determines an environment variable.
+// Variable saves address of installation's app mapping's file.
+// File is given by a bucket name and key delimited by a `/`
+// example: export MM_APPS_MAPPINGS = "apps/latest_mappings"
+const appsMappingsEnvVarName = "MM_APPS_MAPPINGS"
+
+const oldVersionKey = "update_from_version"
+
+// Mappings describes mappings for all the apps in this installation
+// for now it supports lambda functions and static assets
+// appID -> app manifest
+type Mappings struct {
+	Apps map[api.AppID]*api.Manifest
+}
+
+// MappingsFromJSON deserializes a list of app mappings from json.
+func MappingsFromJSON(data []byte) (Mappings, error) {
+	var mappings Mappings
+	if err := json.Unmarshal(data, &mappings); err != nil {
+		return Mappings{}, err
+	}
+	return mappings, nil
+}
+
+// ToJSON serializes a list of app mappings to json.
+func (m *Mappings) ToJSON() []byte {
+	b, _ := json.Marshal(m)
+	return b
+}
+
+// SynchronizeApps synchronizes apps upgrading, downgrading and deleting apps
+func (adm *Admin) SynchronizeApps() error {
+	mappingsFile := os.Getenv(appsMappingsEnvVarName)
+	if mappingsFile == "" {
+		return nil
+	}
+	list := strings.Split(mappingsFile, "/")
+	if len(list) != 2 {
+		return errors.Errorf("Wrong format of an env var - %s", mappingsFile)
+	}
+	bucket := list[0]
+	key := list[1]
+
+	resp, err := adm.awsClient.S3FileDownload(bucket, key)
+	if err != nil {
+		return errors.Wrapf(err, "can't download file %s/%s", bucket, key)
+	}
+
+	mappings, err := MappingsFromJSON(resp)
+	if err != nil {
+		return errors.Wrap(err, "can't deserialize mappings file")
+	}
+
+	listedApps := adm.store.ListApps()
+
+	listedAppsMap := map[api.AppID]*api.App{}
+	updatedAppVersionsMap := map[api.AppID]string{}
+	// Update apps
+	for _, listedApp := range listedApps {
+		listedAppsMap[listedApp.Manifest.AppID] = listedApp
+		manifest, ok := mappings.Apps[listedApp.Manifest.AppID]
+		if !ok {
+			// TODO should deleting the app be that easy?
+			if err := adm.DeleteApp(listedApp); err != nil {
+				return errors.Wrapf(err, "can't delete an app")
+			}
+		}
+		if manifest.Version != listedApp.Manifest.Version {
+			// update app in proxy plugin
+			oldVersion := listedApp.Manifest.Version
+			listedApp.Manifest.Version = manifest.Version
+			if err := adm.store.StoreApp(listedApp); err != nil {
+				return errors.Wrap(err, "can't store app")
+			}
+			updatedAppVersionsMap[listedApp.Manifest.AppID] = oldVersion
+		}
+	}
+
+	// Add new apps as listed
+	for appID, manifest := range mappings.Apps {
+		if _, ok := listedAppsMap[appID]; !ok {
+			if err := adm.AddApp(manifest); err != nil {
+				return errors.Wrap(err, "can't add new app as listed")
+			}
+		}
+	}
+
+	listedAppsUpgraded := adm.store.ListApps()
+
+	// call onInstanceStartup. App migration happens here
```

Comment:
Should we call it onInstanceStartup/OnStartup or OnActivate (to keep in sync with the plugin naming)?

Reply:
I think name was intentionally changed/introduced by Lev not to mess with the plugin's `onAvtivate`. I'm 0/5 on naming. @levb thoughts?

Reply:
Maybe then smth. like OnAppActivate to establish some kind of the parity/continuity? One goal is to reduce the amount of re-thinking for a plugin developer that wants to develop apps.

Reply:
Ditto my previous comment, need to deduplicate for HA (or have the message come from higher-level the Cloud scripts, that start the instance?)

Reply:
This actually depends on our architecture. Should the `OnStartup` function be called for every node of an HA environment? For plugins `onActivate` is called for every node. How about `onDelete`?

Reply:
For plugins this too is an issue, and we are struggling with it (or just plain lucky that the synchronization issues don't bubble up, like in Jira 3.0 data migration)

3/5 We really want cluster-level lifecycle events for Apps, not instance-level.